### PR TITLE
[BIOMAGE-1451] - Dot plot backend

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue

https://biomage.atlassian.net/browse/BIOMAGE-1451

#### Link to staging deployment URL

https://ui-agi-dotplot-final-final.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/546
API - https://github.com/biomage-ltd/api/pull/253
Worker - https://github.com/biomage-ltd/worker/pull/209

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

- Add actions to fetch DotPlot data from the backend

### Code updates
Have best practices and ongoing refactors being observed in this PR

- [x] Migrated any selector / reducer used to the new format
- [x] Validated that current unit tests work as expected and are enough

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [x] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/utils/work/fetchWork.js [hunk 1]
```diff
@@ -143,13 +143,15 @@ const fetchWork = async (
   });
 
   // First, let's try to fetch this information from the local cache.
-  const data = await cache.get(ETag);
-  if (data) {
-    return data;
-  }
+  // const data = await cache.get(ETag);
```
Comment (kafkasl) [inline src/utils/work/fetchWork.js]: why is this commented? should it be removed?

File: src/utils/work/fetchWork.js [hunk 2]
```diff
@@ -143,13 +143,15 @@ const fetchWork = async (
...
+  // if (data) {
+  //   return data;
+  // }
 
   // Then, we may be able to find this in S3.
-  let response = await seekFromS3(ETag);
+  // let response = await seekFromS3(ETag);
```
Comment (kafkasl) [inline src/utils/work/fetchWork.js]: same here

File: src/utils/work/generatePlotWorkBody.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    cellClassKey: config.selectedCellSet,
+    cellSetKey: config.selectedPoints,
+  },
+});
+
+const generatePlotWorkBody = (plotType, config) => {
+  switch (plotType) {
+    case plotNames.plotType.DOT_PLOT:
+      return composeDotPlotWorkBody(config);
+    default: {
```
Comment (kafkasl) [inline src/utils/work/generatePlotWorkBody.js:23]: why this switch? is this something you envision will be re-used by other plots at some point?

File: src/__test__/utils/work/generatePlotWorkBody.test.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import plotNames from 'utils/plots/plotNames';
+import generatePlotWorkBody from 'utils/work/generatePlotWorkBody';
+
+import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
+
```
Comment (kafkasl) [inline src/__test__/utils/work/generatePlotWorkBody.test.js:5]: add ideally two more test cases, one with a wrong plotName, the other with invalid config (to make sure it does not crash the whole UI).

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 5]
```diff
@@ -89,40 +104,81 @@ const DotPlotPage = (props) => {
...
-      const tableState = {
-        pagination: {
-          current: 1, pageSize: config.nMarkerGenes, showSizeChanger: true, total: 0,
-        },
-        geneNamesFilter: null,
-        sorter: { field: PROPERTIES[0], columnKey: PROPERTIES[0], order: 'descend' },
-      };
+  const currentConfig = useRef(null);
+
+  const getCurrentDataProps = (conf) => ({
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: why do we need this function? what extra fields does `conf` contain that are not needed? do they need to be removed?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 6]
```diff
@@ -89,40 +104,81 @@ const DotPlotPage = (props) => {
...
+  });
 
-      dispatch(loadPaginatedGeneProperties(experimentId, PROPERTIES, plotUuid, tableState));
+  const sameConfig = () => _.isEqual(
+    currentConfig.current,
+    getCurrentDataProps(config),
+  );
+
+  useEffect(() => {
+    if (config && !sameConfig()) {
```
Comment (kafkasl) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: try to avoid these kind of functions `sameConfig()` where instead of passing parameters you rely on getting them from the class. It's harder to test and reason about these functions. If you have the function `sameConfig(cfg1, cfg2)` it is very easy to write a test for that. Instead, with `sameConfig()` how do you test it? you need to mock the class to get the `currentConfig.current` and then `config` is not even defined. 

Moreover you are using `getCurrentDataProps(config)` but `config` is not a parameter so you are relying on the variable having the same name in the caller function which is very brittle and will break if someone changes the name.

**Review by kafkasl — DISMISSED** [review]
I don't see any test for the dotPlot component itself. any reason for that?

File: src/utils/work/fetchWork.js (hunk 1 above)
Comment (aerlaut) [reply src/utils/work/fetchWork.js]: Yes, this was commented to bypass cache in development. Uncommented

File: src/__test__/utils/work/generatePlotWorkBody.test.js (hunk 4 above)
Comment (aerlaut) [reply src/__test__/utils/work/generatePlotWorkBody.test.js:5]: Thank you for the input on checking for error. The function will throw an error if no configs match the passed plot name. I've added a test for that.

I don't think we can check for the work request body created by this function, unless we want to have some kind of schema validator like we have in the API. So in the case of invalid wokRequest body, the API will return an error response to the UI (validation error)

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 5 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: This is a helper function to make it easy to get the props that defines the plot's data. The idea is we want the plot to re-fetch the plot's data if these data changes. If other plot config changes, we want the plot to update, but not refetch the data.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 6 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: I understand. I'm removing the wrapper function to remove the dependency.

`getCurrentDataProps` (renamed to `getPlotDataConfig`) actually takes and use a parameter - `conf`. It does not use the component's `config` object. I move the function out of the component body, so it's clearer.

File: src/utils/work/generatePlotWorkBody.js (hunk 3 above)
Comment (aerlaut) [reply src/utils/work/generatePlotWorkBody.js:23]: Yes. Other plots calculate plot data from data that is already available on the platform (e.g. gene expression), so they do not need to define their own work request body. This is the first plot that has its own work request to specifically calculate data for the plot. I think this functionality will be needed for new plots in the platform.

**Comment by aerlaut** [comment]
> I don't see any test for the dotPlot component itself. any reason for that?

This test is already implemented in `DotPlot.test.jsx`, it was added in PR #529. I think it has tested for cases that are available in `DotPlot.jsx`.

File: src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx [hunk 7]
```diff
@@ -55,6 +55,7 @@ describe('MarkerGeneSelection', () => {
 
   it('Should show the number of marker genes input', async () => {
     const mockConfig = _.merge(
+      {},
```
Comment (kafkasl) [inline src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx]: what's the point of merging something empty? is the whole `_.merge` is needed at all?

File: src/__test__/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.test.jsx [hunk 8]
```diff
@@ -99,4 +123,26 @@ describe('Dot plot page', () => {
...
+
+    await act(async () => {
+      render(
+        <Provider store={storeState}>
+          {dotPlotPageFactory()}
+        </Provider>,
+      );
+    });
+
+    expect(screen.getByText(/Error loading cell sets/i)).toBeInTheDocument();
```
Comment (kafkasl) [inline src/__test__/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.test.jsx:157]: for this test file have you tried what we discussed about checking that tests are doing something? e.g. pass a correct config response in the last one and expect it to get an error (so making sure it fails), and so on

File: src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx (hunk 7 above)
Comment (aerlaut) [reply src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx]: `_.merge` modifies the first object that's passed to it, and also returning the modified object. If i do `_.merge(a, b)`, the object `a` is modified with changes in `b`. If I do `_.merge({}, a, b)`, `a` is not modified, and the function returns a new object that is a combination of `a` and `b`.

In this case, I don't want to modify the `initialPlotConfigStates.dotPlot`, but I want an object that contains `initialPlotConfigStates.dotPlot` and `{ useCustomGenes: false }`. It is like doing 
      
```
{
   ...initialPlotConfigStates.dotPlot,
   useCustomGenes: false
}
```

File: src/__test__/redux/reducers/componentConfig/loadPlotData.test.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  it('Throws an error if there is an error fetching plot data', async () => {
+    fetchWork.mockImplementation(() => {
+      throw new Error('Error fetching work');
+    });
+
+    await act((async () => {
+      await storeState.dispatch(loadPlotData(experimentId, plotUuid, plotType));
+    }));
+
+    expect(pushNotificationMessage).toHaveBeenCalled();
```
Comment (kafkasl) [inline src/__test__/redux/reducers/componentConfig/loadPlotData.test.js]: 1. add how many times do you expect the function to be called
2. is checking if the function has been called enough to test that `throws an error if there is an error fetching plot data`?

File: src/redux/reducers/componentConfig/initialState.js [hunk 10]
```diff
@@ -754,7 +754,18 @@ const initialPlotConfigStates = {
 const initialComponentConfigStates = {
   interactiveHeatmap: interactiveHeatmapInitialConfig,
 };
-export { initialPlotConfigStates, initialComponentConfigStates };
+
+const initialPlotDataState = {
```
Comment (kafkasl) [inline src/redux/reducers/componentConfig/initialState.js:774]: where was the initial plot data defined before? is this supposed to be reused (is it already?) by other plots? does it match the other plots' initial data state?

File: src/redux/reducers/componentConfig/plotDataLoading.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { initialPlotDataState } from './initialState';
+
+const plotDataLoading = (state, action) => {
```
Comment (kafkasl) [inline src/redux/reducers/componentConfig/plotDataLoading.js:1]: why are we storing this? to show some placeholder while we load? what did we do before that?

File: src/utils/plots/plotNames.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+const plotType = {
```
Comment (kafkasl) [inline src/utils/plots/plotNames.js]: this contains more than plot names, I would name the file `contants.js`. It can stay in the plots folder but I believe it's best to place it directly inside the utils folder because otherwise it's easier to miss it, and the next developer might add his contacts somewhere else

File: src/utils/work/generatePlotWorkBody.js (hunk 3 above)
Comment (kafkasl) [reply src/utils/work/generatePlotWorkBody.js:23]: in this situation it probably makes sense as I can foresee having a new plot that needs worker data. However, be very wary of coding for features not needed because it just adds extra complexity that might not be needed or used in the end. https://en.wikipedia.org/wiki/KISS_principle

File: src/__test__/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.test.jsx (hunk 8 above)
Comment (aerlaut) [reply src/__test__/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.test.jsx:157]: yes, I've made sure the tests fail if it's passed the correct response

File: src/redux/reducers/componentConfig/plotDataLoading.js (hunk 11 above)
Comment (aerlaut) [reply src/redux/reducers/componentConfig/plotDataLoading.js:1]: Yes, to show placeholder while the plot is being loaded. Before this, we are checking for `!plotData.length` to show that the item is loading. I think having this `loading` state is much better, because we might have cases where plot data is empty (there actually is no data) while we are not loading anymore.

File: src/redux/reducers/componentConfig/initialState.js (hunk 10 above)
Comment (aerlaut) [reply src/redux/reducers/componentConfig/initialState.js:774]: `plotData` was initially added as the payload in `loadPlotConfig` when data is being fetched from the server. `plotData` is used to pass data from the server to the plots. Not all plots have `plotData`. Right now, only plots in the filters in `data-processing` uses `plotData`.

File: src/__test__/redux/reducers/componentConfig/loadPlotData.test.js (hunk 9 above)
Comment (aerlaut) [reply src/__test__/redux/reducers/componentConfig/loadPlotData.test.js]: Whoops, it should display notification, not throw an error.

File: src/redux/reducers/componentConfig/initialState.js (hunk 10 above)
Comment (aerlaut) [reply src/redux/reducers/componentConfig/initialState.js:774]: I have modified `loadPlotConfig` to make it. I have also inserted `initialPlotDataState` when loading plot config. This way, all plots have the same states, although they might not use it.

File: src/__test__/redux/reducers/componentConfig/loadPlotData.test.js (hunk 9 above)
Comment (aerlaut) [reply src/__test__/redux/reducers/componentConfig/loadPlotData.test.js]: I've also modified `toHaveBeenCalled()` into `toHaveBeenCalledTimes(1)`

File: src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx (hunk 7 above)
Comment (kafkasl) [reply src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx]: and why you didn't use? it looks cleaner and easier to read, is there any drawback?
```
{ ...initialPlotConfigStates.dotPlot, useCustomGenes: false }
```

Comment (aerlaut) [reply src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx]: `_.merge()` is useful when merging objects with multiple layers. In this case it isn't, so it's over the top. I'll change it to use object spreading.

Comment (aerlaut) [reply src/__test__/components/plots/styling/MarkerGeneSelection.test.jsx]: Changed to use simpler merge

File: src/components/plots/styling/MarkerGeneSelection.jsx [hunk 13]
```diff
@@ -15,62 +15,62 @@ const MarkerGeneSelection = (props) => {
   const [numGenes, setNumGenes] = useState(config.nMarkerGenes);
 
   const renderOptions = () => {
-    if (config.useCustomGenes) {
```
Comment (aerlaut) [inline src/components/plots/styling/MarkerGeneSelection.jsx:18]: This is just switching to make it easier to read

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+      await testStore.dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
+    });
+
+    // Expect componentConfig to contain key for plotUuid
+    expect(testStore.getState().componentConfig[plotUuid]).toBeDefined();
+
+    const plotConfig = testStore.getState().componentConfig[plotUuid];
+
+    expect(plotConfig.config).toBeDefined();
+    expect(plotConfig.loading).toBeDefined();
```
Comment (kafkasl) [inline src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js]: why not just check the actual value instead of `toBeDefined` ?

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    expect(plotConfig.error).toBeDefined();
+    expect(plotConfig.plotType).toBe(plotType);
+
+    expect(plotConfig.plotData).toBeDefined();
+    expect(_.isEqual(plotConfig.plotData, mockData.plotData)).toBe(true);
+
+    expect(plotConfig).toMatchSnapshot();
```
Comment (kafkasl) [inline src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js:52]: if you are matching the snapshot here, why add all those checks before this line which are essentially doing the same?

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+    expect(plotConfig.config).toBeDefined();
+    expect(plotConfig.loading).toBeDefined();
+    expect(plotConfig.error).toBeDefined();
+    expect(plotConfig.plotType).toBe(plotType);
+
+    expect(plotConfig.plotData).toBeDefined();
+    expect(plotConfig.plotData.length).toEqual(0);
+
+    expect(plotConfig).toMatchSnapshot();
```
Comment (kafkasl) [inline src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js:67]: same here, either check the snapshot or individual fields, I don't see why would you do both

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js (hunk 14 above)
Comment (aerlaut) [reply src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js]: This code is removed, checking of the properties is done with `toMatchSnapshot()`

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js (hunk 15 above)
Comment (aerlaut) [reply src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js:52]: Use this instead of manually checking the properties, which is not robust

File: src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js (hunk 16 above)
Comment (aerlaut) [reply src/__test__/redux/actions/componentConfig/loadPlotConfig.test.js:67]: This code is removed, checking of the properties is done with `toMatchSnapshot()`

File: src/__test__/redux/reducers/componentConfig/loadPlotData.test.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  )),
+}));
+
+const experimentId = fake.EXPERIMENT_ID;
+let testStore = null;
+
+const plotUuid = 'DotPlotMain';
+const plotType = plotTypes.DOT_PLOT;
+
+describe('plotDataLoaded', () => {
```
Comment (aerlaut) [inline src/__test__/redux/reducers/componentConfig/loadPlotData.test.js]: `loadPlotData` is changed to `fetchPlotDataWork`. Change the test name accordingly

File: src/redux/actions/componentConfig/fetchPlotDataWork.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
+import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
+import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
+import { PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR } from 'redux/actionTypes/componentConfig';
+
+import pushNotificationMessage from 'utils/pushNotificationMessage';
+import endUserMessages from 'utils/endUserMessages';
+import generatePlotWorkBody from 'utils/work/generatePlotWorkBody';
+import { fetchWork } from 'utils/work/fetchWork';
+
+const fetchWorkerPlotData = (experimentId, plotUuid, plotType) => async (dispatch, getState) => {
```
Comment (aerlaut) [inline src/redux/actions/componentConfig/fetchPlotDataWork.js]: Change the name to be the same as the file name

File: src/components/plots/DotPlot.jsx [hunk 19]
```diff
@@ -39,7 +20,7 @@ const DotPlot = (props) => {
     export: true,
     source: false,
     compiled: false,
-    editor: false,
+    editor: true,
```
Comment (aerlaut) [inline src/components/plots/DotPlot.jsx]: change editor to false

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 20]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
-  const cellSets = useSelector((state) => state.cellSets);
+
+  const cellSets = useSelector(getCellSets());
   const {
     loading: cellSetsLoading,
     error: cellSetsError,
-    hierarchy,
+    hierarchy: cellSetHierarcy,
+    properties: cellSetProperties,
   } = cellSets;
```
Comment (aerlaut) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:111]: Use selector directly on line 111

File: src/redux/reducers/componentConfig/plotDataLoading.js [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+  const { plotUuid } = action.payload;
+  return {
+    ...state,
+    [plotUuid]: {
+      ...initialPlotDataState,
```
Comment (aerlaut) [inline src/redux/reducers/componentConfig/plotDataLoading.js]: Check if we still need this

File: src/utils/work/fetchWork.js [hunk 22]
```diff
@@ -119,7 +119,7 @@ const fetchWork = async (
   }
 
   // If caching is disabled, we add an additional randomized key to the hash so we never reuse
-  // past results.
+  // past results. Use the string 'true' because localStorage only store values as strings
```
Comment (aerlaut) [inline src/utils/work/fetchWork.js]: Moved to another PR

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 20 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:111]: changed

File: src/components/plots/DotPlot.jsx (hunk 19 above)
Comment (aerlaut) [reply src/components/plots/DotPlot.jsx]: changed

File: src/__test__/redux/reducers/componentConfig/loadPlotData.test.js (hunk 17 above)
Comment (aerlaut) [reply src/__test__/redux/reducers/componentConfig/loadPlotData.test.js]: changed

File: src/redux/actions/componentConfig/fetchPlotDataWork.js (hunk 18 above)
Comment (aerlaut) [reply src/redux/actions/componentConfig/fetchPlotDataWork.js]: changed

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 23]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
+
+    // filterBy has the shape louvain/louvain-1
+    const [filterRootNode, filterKey] = filterCluster.split('/');
+
+    // If 'All" is chosen for the dropdown, there will always be representation from more than 1 group
+    if (filterRootNode === 'All') return true;
+
+    const filterClusterCellIds = Array.from(cellSetProperties[filterKey].cellIds);
+
+    const baseClusterParent = cellSetHierarcy.find((cellSet) => cellSet.key === baseCluster);
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: Check if you can use the cellsets selectors here I think we can get hierarchy by key.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 24]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
+    return baseClusterCellIds.some((baseCellIds) => {
+      if (filterClusterCellIds.some((filterCellId) => baseCellIds.has(filterCellId))) {
+        numIntersections += 1;
+      }
+
+      return numIntersections === 2;
+    });
+  };
+
+  const hasGroups = useMemo(
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: Isnt it better to use `useCallback` here?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 25]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
-        },
-        geneNamesFilter: null,
-        sorter: { field: PROPERTIES[0], columnKey: PROPERTIES[0], order: 'descend' },
-      };
+    if (cellSetsLoading) return;
 
-      dispatch(loadPaginatedGeneProperties(experimentId, PROPERTIES, plotUuid, tableState));
+    // Marker genes calculation needs that the cellIds in groupBy (refer to fn definition)
+    // be represented by more than one groups in filterBy to enable comparison
+    // if (config?.useMarkerGenes && !hasGroupsToCompare(config.selectedCellSet, config.selectedPoints)) {
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: remove this

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 26]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
+      if (filterClusterCellIds.some((filterCellId) => baseCellIds.has(filterCellId))) {
+        numIntersections += 1;
+      }
+
+      return numIntersections === 2;
+    });
+  };
+
+  const hasGroups = useMemo(
+    () => hasGroupsToCompare(config?.selectedCellSet, config?.selectedPoints),
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: Initially when the config is not defined we still do this calculation in the function, maybe its better to put a check in `hasGroupsToCompare` if config is defined?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 24 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: No, because the return value is a boolean, not a function. `useCallback` memoizes a function.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 27]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
+      },
+      geneNamesFilter: null,
+      sorter: { field: PROPERTIES[0], columnKey: PROPERTIES[0], order: 'descend' },
+    };
+
+    dispatch(loadPaginatedGeneProperties(experimentId, PROPERTIES, plotUuid, tableState));
+  };
 
   useEffect(() => {
+    if (config?.selectedGenes.length === 0 && !highestDispersionGenes?.length > 0 && !genesFetching) {
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:198]: What is this expression - `!highestDispersionGenes?.length > 0`? maybe you mean `!highestDispersionGenes.length`

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 25 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: The comment? I think it's good information so that people knows why we need to check for this,

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 26 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: There is already a check inside `hasGroupsToCompare` to check for the existence of the inputs. I'll move it out into the `useMemo` function body

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx [hunk 28]
```diff
@@ -71,75 +89,249 @@ const DotPlotPage = (props) => {
...
+    setMoreThanTwoGroups(true);
+
+    const currentComparedConfig = getComparedConfig(config);
+    if (config && !_.isEqual(previousComparedConfig.current, currentComparedConfig)) {
+      previousComparedConfig.current = currentComparedConfig;
+      dispatch(fetchPlotDataWork(experimentId, plotUuid, plotType));
+    }
+  }, [config, cellSetProperties]);
+
+  const loadInitialCustomGenes = () => {
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:184]: Other plots use this function such as the violin plot and embedding continuous, can you check if you can move this function and reuse it in those places?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 27 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:198]: We want this check to pass if `highestDispersionGene` is not loaded.

`!highestDispersionGenes?.length > 0` is a combination of 2 checks : 
-if `highestDispersionGenes` is `undefined`, then it evaluates to `!undefined` which is true
-if `heighestDispersionGenes` is defined and the data is still loading (`length` is 0), then it evaluates to `!([].length > 0)` will evaluate to `true`

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 28 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:184]: I don't want to do that refactoring in this PR because it's too big already. I can make another ticket for it. Sounds good?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 25 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx]: Oh the commented line of code. Good catch! removed.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 27 above)
Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:198]: yeah but !highestDispersionGenes?.length will give the same output in those cases, its just a bit confusing when you see it at first maybe its me

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 23 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: Whoops, you've just uncovered a bug in React! I tried using a selector in the function (which is called inside a `useMemo`). Looking at the error, I don' think we can do this yet.

image

Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: no its not good to use hook inside a hook in general i think its one of the react hook rules, maybe you can do it on the top of the file as a variable.

Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: Also alternatively you can call the actual selector directly not through useSelector i guess?

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 27 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:198]: Yup, which is what we want. I just checked `loadPaginatedGeneProperties`. If `highestDispersionGenes` is undefined (first call), it is the call to initialize the state. On error, it populates `data` to an empty array. So we also want to be able to re-dispatch this call if there is a previous error (`data` is empty).

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 23 above)
Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: I don't think we can do that, because `getCellSetsSelector` requires `state`, which is passed by `useSelector`.

Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: I can move that line outside the function, but it makes it messy because we get the `baseClusterParent` as another global dependency from outside of the function and we lose the use of `baseCluster` as one of the keys of memoization.

Using the selectors is a good thing - the result is memoized. The result of the `hasGroups...` function is also memoized and the cellSetsHierarchy array is short, so I don't think the performance hit of not using the memoized selector would be a noticeable hit.

Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: This works,  but as I said, we lose the functionality of `baseCluster` as a key for useMemo, and we can't use `baseClusterParent` as a key as it is an object (whose reference changes everytime)

image

Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: So I'd still use the current approach. wdyt?

File: src/redux/reducers/componentConfig/plotDataLoading.js (hunk 21 above)
Comment (aerlaut) [reply src/redux/reducers/componentConfig/plotDataLoading.js]: not needed anymore as it is initially initialized when loading the plot config, so this is removed.

File: src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx (hunk 23 above)
Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: You can pass the hierarchy and keys directly and  call `getCellSetsHierarchyByKeys`? As for the other things im not sure if it doesnt impact performance then maybe its good?

Comment (aerlaut) [reply src/pages/experiments/[experimentId]/plots-and-tables/dot-plot/index.jsx:142]: Yup, that's what I did, as I show in the example above. However, I didn't really like it because : 
- It splits the code so that `baseClusterParent` away from where it's used, and causes `hasGroupsToCompare` to have another dependency on a global variable. 
- The data that is fetched by `getCellSetsHierarchyByKeys` already exists in `cellSets`, which is used in other parts of the code.
- Performance-wise, it's not that much of a difference. So I think the current code (without the selector) looks better as it has better legibility.

So I think the way it's currently coded looks fine.

**Review by StefanBabukov — APPROVED** [review]
